File: src/main/java/shootemup/Hitbox.java
package shootemup;

import java.util.List;
import java.util.stream.IntStream;

import shootemup.geometry.IRange;
import shootemup.geometry.Vector;


public class Hitbox {

    public Hitbox(List<IRange> dimensions) {
	assertEnough(dimensions);
	this.dimensions = dimensions;
    }

    public boolean contains(Vector point) {
	for (int i = 0; i < dimensions.size(); i++) {
	    IRange dimension = dimensions.get(i);
            Integer element = point.elements.get(i);
            if (!dimension.contains(element))
		return false;
	}
	return true;
    }

    private void assertEnough(List<IRange> dimensions) {
        if (dimensions.size() <= 0)
	    throw new IllegalArgumentException("Hitbox cannot be 0-dimensional");
    }

    private final List<IRange> dimensions;

}

File: src/main/kotlin/shootemup/Hitbox.kt
package shootemup

import shootemup.geometry.*

class Hitbox(private val dimensions: Dimensions) {

    init { checkNumberOf(dimensions) }

    fun contains(point: Vector): Boolean {
        for (i in dimensions.indices) {
            val dimension = dimensions.get(i)
            val element = point.get(i)
            if (!dimension.contains(element))
                return false
        }
        return true
    }

    private fun checkNumberOf(dimensions: Dimensions) {
        if (dimensions.size <= 0)
            throw IllegalArgumentException("Hitbox cannot be 0-dimensional")
    }

}
